index.js

```diff
@@ -184,7 +235,21 @@ function registerPlugin (fastify, meta, allPlugins, parentPlugins = {}) {
     }
   }
 
-  fastify.register(plugin, options)
+  if (hooks && hooks.length > 0) {
+    // plugin has hooks, which we need to encapsulate at the same layer as the plugin
+    fastify.register(async function (app, opts) {
+      for (const hook of hooks) {
+        if (hook.default) hook.default[Symbol.for('skip-override')] = true
+        app.register(hook, options)
+      };
+
+      app.register(plugin, options)
+    })
```

Comment:
I think we should invert this relationship as it can be inefficient as it would load them for all files. I recommend we create a new plugin, load all hooks in a folder, and then we call `registerPlugin` for all the routes files.

Reply:
Sounds good, I moved that logic to a wrapper function and now pass in a completed plugin object for the `registerPlugin` call. Does that make more sense to you?

Reply:
That's not what I'm looking for close.

In https://github.com/fastify/fastify-autoload/blob/fa95893d3d5fc954bece8ffd8d3cd974d7154d78/index.js#L43, we area loading all plugins in a loop. We are flattening a tree into a list. This will cause us executing the same hooks for each file in the in the given sub-tree.

The most efficient way of loading this would be to retain the information and just load them once (in pseudocode):

```js
    // plugin has hooks, which we need to encapsulate at the same layer as the plugin
    fastify.register(async function (app, opts) {
      app.register(hooks)
      for (const plugin of pluginsInThisFolder) {
        app.register(plugin, options)
      }
    })
```

Reply:
That makes a lot of sense, I will take a look tonight!